x-pack/plugins/endpoint/server/routes/endpoints.test.ts

```diff
@@ -0,0 +1,183 @@
+/*
```

Comment:
Would this test be better as an API integration test? It could load an archvie (via es_archiver) and instrument the route handler and ES. This would check that the routing works and that the query works. Also it wouldn't require writing any mocking.

Reply:
> Would this test be better as an API integration test? It could load an archvie (via es_archiver) and instrument the route handler and ES. This would check that the routing works and that the query works. Also it wouldn't require writing any mocking.

We still need a unit test. We have an api test too but I did not want to add it to this PR to blow it up. I can add it to another PR. But I can add it if you insist.

Reply:
Can you explain why you still need a unit tests? My understanding is that this test mocks various internals and asserts that they were called. My concern is that this type of test doesn't check the behavior. I'm also concerned that since its completely coupled to internals, it might break even if the behavior is correct.

Reply:
> Can you explain why you still need a unit tests? My understanding is that this test mocks various internals and asserts that they were called

This is what unit tests do, they test internals and expected calls on dependencies with stubs. We cannot defer all tests tills the after the entire system is coupled together and started, without covering the units.

Reply:
Thanks for adding the integration test. I agree that mocking can cause tight coupling and personally regret the places I did it in the SMP. I can see the value in stubbing out ES so you can test the translation from the search results of endpoints into the returned format done by this function `mapToEndpointResultList`. My suggestion (only a suggestion) would be to leave out the expects for how a function is called and how many times it is called. Like these lines:
`expect(mockScopedClient.callAsCurrentUser).toBeCalledWith('search', {`
`const endpointResultList = mockResponse.ok.mock.calls[0][0]?.body as EndpointResultList;`

and instead just validate that the results get transformed into the expected format like you're already doing.

Reply:
Based on @jonathan-buttner's comment, I'm wondering if this could be replaced by one (or more) unit tests against a pure function called `mapToEndpointResultList`. This function could be responsible for any intricate business logic and could be tested without stubs/mocking.

A second set of API integration tests could instrument the vertical slice but with much less coverage. This approach might allow your tests to be less coupled to the implementation while also having no mocking/stubs.

I'm also arguing that unit tests don't necessarily require mocking. Here is the PR I'm working on: https://github.com/elastic/kibana/pull/53619/files

Here are my unit test files. These do not use mocks/stubs and they also test the behavior of a small unit:

* [x-pack/plugins/endpoint/public/embeddables/resolver/lib/matrix3.ts](https://github.com/elastic/kibana/pull/53619/files#diff-8cb75023583cda940ccc013c5df3bdcc)
* [x-pack/plugins/endpoint/public/embeddables/resolver/lib/transformation.test.ts](https://github.com/elastic/kibana/pull/53619/files#diff-532e4d475d8fd0e891297c3a34fae8e9)
* [x-pack/plugins/endpoint/public/embeddables/resolver/store/camera/inverse_projection_matrix.test.ts](https://github.com/elastic/kibana/pull/53619/files#diff-3fd70733fa0c53321aefd73369b09360)
* [x-pack/plugins/endpoint/public/embeddables/resolver/store/camera/panning.test.ts](https://github.com/elastic/kibana/pull/53619/files#diff-7c5f2e50f8d001a64a2e5415533a5009)
* [x-pack/plugins/endpoint/public/embeddables/resolver/store/camera/projection_matrix.test.ts](https://github.com/elastic/kibana/pull/53619/files#diff-cd916893827887afe721dbf3cccb7121)
* [x-pack/plugins/endpoint/public/embeddables/resolver/store/camera/zooming.test.ts](https://github.com/elastic/kibana/pull/53619/files#diff-5d9de62597219aed4af527874e99cf47)

My integration / functional tests will instrument all of the 'plumbing' that is responsible for setting up react, redux, etc. Those tests will have less coverage (they will no instrument all the branches instrumented by the unit tests.)